trin-core/src/portalnet/overlay.rs

```diff
@@ -0,0 +1,169 @@
+    kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+}
+
+impl Overlay {
+    pub fn new(local_enr: Enr, data_radius: U256, config: Config) -> Self {
+        let kbuckets = Arc::new(RwLock::new(KBucketsTable::new(
+            local_enr.node_id().into(),
+            config.bucket_pending_timeout,
+            config.max_incoming_per_bucket,
+            config.table_filter,
+            config.bucket_filter,
+        )));
+        let local_enr = Arc::new(RwLock::new(local_enr));
+        let data_radius = Arc::new(RwLock::new(data_radius));
+
+        Self {
+            local_enr,
+            data_radius,
+            kbuckets,
+        }
+    }
+
+    /// Returns the local ENR of the node.
+    pub fn local_enr(&self) -> Enr {
+        self.local_enr.read().clone()
+    }
+
+    // Returns the data radius of the node.
+    pub fn data_radius(&self) -> U256 {
+        self.data_radius.read().clone()
+    }
+
+    /// Returns a vector of the ENRs of the closest nodes by the given distances.
+    pub fn nodes_by_distance(&self, mut distances: Vec<u64>) -> Vec<Enr> {
+        let mut nodes_to_send = Vec::new();
+        distances.sort_unstable();
+        distances.dedup();
+
+        if let Some(0) = distances.first() {
+            // if the distance is 0 send our local ENR
+            nodes_to_send.push(self.local_enr.read().clone());
+            distances.remove(0);
+        }
+
+        if !distances.is_empty() {
+            let mut kbuckets = self.kbuckets.write();
+            for node in kbuckets
+                .nodes_by_distances(distances, FIND_NODES_MAX_NODES)
+                .into_iter()
+                .map(|entry| entry.node.value.clone())
+            {
+                nodes_to_send.push(node.enr());
+            }
+        }
+        nodes_to_send
+    }
+
+    /// Returns list of nodes closer to content than self, sorted by distance.
+    pub fn find_nodes_close_to_content(&self, content_key: Vec<u8>) -> Vec<SszEnr> {
+        let self_node_id = self.local_enr().node_id();
+        let self_distance = xor_two_values(&content_key, &self_node_id.raw().to_vec());
+
+        let mut nodes_with_distance: Vec<(Vec<u8>, Enr)> = self
+            .table_entries_enr()
+            .into_iter()
+            .map(|enr| {
+                (
+                    xor_two_values(&content_key, &enr.node_id().raw().to_vec()),
+                    enr,
+                )
+            })
+            .collect();
+
+        nodes_with_distance.sort_by(|a, b| a.0.cmp(&b.0));
+
+        let closest_nodes = nodes_with_distance
+            .into_iter()
+            .take(FIND_CONTENT_MAX_NODES)
+            .filter(|node_record| &node_record.0 < &self_distance)
+            .map(|node_record| SszEnr::new(node_record.1))
+            .collect();
+
+        closest_nodes
+    }
+
+    /// Returns an iterator over all ENR node IDs of nodes currently contained in the routing table.
+    pub fn table_entries_id(&self) -> Vec<NodeId> {
```

Comment:
Not sure, but I think it makes sense to combine this and the next method `tables_entries_enr` into one and return a vector of tuples.